notifier/callback.go

```diff
@@ -12,40 +13,39 @@ import (
 // and receive notifications.
 type Callback struct {
 	NotificationID uuid.UUID `json:"notification_id"`
-	Callback       url.URL   `json:"callback"`
+	Callback       url.URL   `json:"callback,string"`
 }
 
+// CbMessage is the actual type used for over the wire. The above's struct tags
+// are for documentation only.
+type cbMessage struct {
+	ID       *uuid.UUID `json:"notification_id"`
+	Callback string     `json:"callback"`
+}
+
+// MarshalJSON implements json.Marshaler.
 func (cb Callback) MarshalJSON() ([]byte, error) {
-	var m = map[string]string{
-		"notification_id": cb.NotificationID.String(),
-		"callback":        cb.Callback.String(),
-	}
-	return json.Marshal(m)
+	// This writes out a block of bytes directly, instead of allocating a
+	// cbMessage structure, serializing the URL, then copying it into a []byte.
+	b := bytes.NewBufferString(`{"callback":"`)
```

Comment:
this looks like a rough thing to maintain if the schema changes. 
is this strictly necessary?

Reply:
Strictly necessary, no, but these methods are the only places where we don't get the piecewise encoding of the `codec` package "for free," so I thought it'd be prudent to minimize the number of intermediary allocations here.

Reply:
Do you think you loose the majority of this PRs effectiveness if we moved away from writing out the json directly?

Reply:
The majority? No, it just seemed like low-hanging fruit.

Reply:
Id prefer to not do this, string building json json is pretty ugly, and not great to maintain. As long as we increase the majority of the efficiency, we can still keeo maintainability in mind. Amdahl's law comes to mind.